style_guide.md

```diff
@@ -0,0 +1,270 @@
+
+Conversely, the risk of an internal fn being overridden and causing a problem is the responsibility of the code author.  
+Additionally, we should strive to design contracts that are flexible enough for someone to import it directly and use it as they see fit. 
+
+
+### calldata versus memory
+For functions that would be called externally, use calldata over memory when declaring parameters(string, bytes) in function definition.
+* Saves gas as calldata is cheaper than memory, and a duplicate is not created. 
+* This is contingent on that fact that the parameter remains immutable throughout the function.
+
+> Use memory if you want to be able to manipulate the parameters, and calldata when the parameter remains immutable.
+
+**Explanation:**
+
+calldata contains parameters of a function as allocated by the external caller.
+Therefore, in external calls to functions (when passing parameter type string), opt to use `calldata string`, instead of `memory string`.
+* using calldata -> will simply reference the pre-allocated memory location
+* using memory -> will create a copy of the parameter in a new memory location and then pass it into the function. 
+
+**Yes:**
+
+When using `(string calldata name)`, the parameter would be passed directly into `delete holder[name]`, without making a copy, thereby saving gas.
+```solidity
+    function release(string calldata name) public {
+        require(holder[name] == msg.sender, "Not your name!");
+        delete holder[name];
+        emit Release(msg.sender, name);
+    }
+```
+
+**No:**
+
+When using `(string memory name)`, a copy of the parameter is passed into `delete holder[name]`:
+```solidity
+    function release(string memory name) public {
+        require(holder[name] == msg.sender, "Not your name!");
+        delete holder[name];
+        emit Release(msg.sender, name);
+    }
+```
+
+**Gotcha!** 
+
+If you pass in a string (or bytes) from a different internal function where you had just created that string in memory, then you can't use calldata. 
+```solidity
+contract X {
+ function sendString() internal {
+    string memory y = "Raja is cool";
+    callFunction(y);
+ }
+
+//ERROR: because calldata is only used from external calls, and y was created in memory before
+ function callFunction(string calldata y) {}  
+}
+```
+* execution fails on `callFunction(y)` in `sendString()`, it is not an external call; there is no msg.data than holds calldata to be passed into callFunction().
+* in the absence of msg.data, there is no calldata that can be accessed. 
+
+
+
+### Caching sload into mload
+Repeated calls to storage variables within the same function should be avoided, if we are only reading its value.
```

Comment:
feels like we might want to have a separate section for gas opt

Reply:
yah I was going back and forth on this. this could very well evolve into a separate standalone doc on best practices solely for gas optimizations. which i think would be very handy reference, not just for us, but perhaps the community at large as well.